subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppCachingIntegrationTest.groovy

```diff
@@ -51,29 +52,116 @@ class CppCachingIntegrationTest extends AbstractCppInstalledToolChainIntegration
     }
 
     def 'compilation can be cached'() {
-        setupProjectInDirectory(temporaryFolder.testDirectory)
+        def f = new NativeCachingFixture(buildType: buildType)
+        setupProject()
 
         when:
-        withBuildCache().succeeds compilationTask
+        withBuildCache().run f.compilationTask
 
         then:
-        compileIsNotCached()
+        f.compileIsNotCached()
 
         when:
-        withBuildCache().succeeds 'clean', ':installDebug'
+        withBuildCache().run 'clean', installTask(buildType)
 
         then:
-        compileIsCached()
-        installation("build/install/main/debug").exec().out == app.expectedOutput
+        f.compileIsCached()
+        installation("build/install/main/${buildType.toLowerCase()}").exec().out == app.expectedOutput
+
+        where:
+        buildType << [debug, release]
+    }
+
+    def "compilation task is relocatable"() {
+        def f = new NativeCachingFixture(buildType: buildType)
+        def originalLocation = file('original-location')
+        def newLocation = file('new-location')
+        setupProject(originalLocation)
+        setupProject(newLocation)
+
+        when:
+        inDirectory(originalLocation)
+        withBuildCache().run f.compilationTask
+
+        def snapshotsInOriginalLocation = f.snapshotObjects(originalLocation)
+
+        then:
+        executedAndNotSkipped f.compilationTask
+
+        when:
+        executer.beforeExecute {
+            usingProjectDirectory(newLocation)
+        }
+        run f.compilationTask
+
+        then:
+        executedAndNotSkipped(f.compilationTask)
+        f.assertSameSnapshots(snapshotsInOriginalLocation, f.snapshotObjects(newLocation))
+
+        when:
+        run 'clean'
+        withBuildCache().run f.compilationTask, installTask(buildType)
+
+        then:
+        skipped f.compilationTask
+        f.assertSameSnapshots(snapshotsInOriginalLocation, f.snapshotObjects(newLocation))
+        installation(newLocation.file("build/install/main/${buildType.toLowerCase()}")).exec().out == app.expectedOutput
+
+        where:
+        buildType << [debug, release]
     }
 
-    void compileIsCached() {
-        assert skippedTasks.contains(compilationTask)
-        objectFileFor(file('src/main/cpp/main.cpp'), "build/obj/main/debug").assertExists()
+    String getSourceType() {
+        return 'Cpp'
     }
 
-    void compileIsNotCached() {
-        assert !skippedTasks.contains(compilationTask)
+    class NativeCachingFixture {
```

Comment:
I find it hard to decipher what is happening here. Would there be a more descriptive name for this type perhaps? What is the purpose of putting these methods in a class?